### PR TITLE
[SPARK-12941][SQL][BRANCH-1.4] Spark-SQL JDBC Oracle dialect fails to map string datatypes to Oracle VARCHAR datatype

### PR DESCRIPTION
**Pull request by thomastechs**
Adding a getJDBCType() method to the JdbcDialects.scala which would create a VARCHAR type for Oracle create table operations.Currently the type TEXT is throwing exceptions which is incompatible to Oracle

### CODE REVIEW

**Comment by thomastechs** [comment]
@yhuai     fyr

**Comment by yhuai** [comment]
Can you also create a pr for master? 

btw, it is possible to add a docker integration test (like MySQLIntegrationSuite)?

**Comment by yhuai** [comment]
test this please

**Comment by yhuai** [comment]
please fix the style error

**Comment by thomastechs** [comment]
@yhuai  The style check error has been fixed and committed.

**Comment by yhuai** [comment]
Can you add a test? Also, can you create a pr for master?

**Comment by thomastechs** [comment]
@yhuai  Added the testcase for 1.4.     For the master PR, would it be good to raise another bug in reference to this and upload the fix as a new PR for that bug.? In that case, we could close this quickly (once tests passed) , and release this fix for Jose who raised this bug, for his purpose.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 1]
```diff
@@ -445,4 +445,9 @@ class JDBCSuite extends SparkFunSuite with BeforeAndAfter {
     assert(agg.getCatalystType(1, "", 1, null) == Some(StringType))
   }
 
+  test("OracleDialect type mapping") {
+    val oracleDialect = JdbcDialects.get("jdbc:oracle://127.0.0.1/db")
+    assert(oracleDialect.getJDBCType(StringType).map(_.databaseTypeDefinition).get == "VARCHAR2(255)")
+  }
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: I am not sure this test works. Let me be clear. We need to add a docker integration test like `MySQLIntegrationSuite`. Since spark 1.4 branch does not have the testing infrastructure. I think the the right way to do it is for your new PR (using the same JIRA id in the title) created for master, you add the test. Then, for this PR, we just keep the fix.

**Comment by yhuai** [comment]
Please create a PR against master. You do not need to create a new jira. In the title of the new PR, you just use the same JIRA number. For this PR, it is better to add `[BRANCH-1.4]` in the title.

**Comment by yhuai** [comment]
btw, just want to double check. Have you tested your fix with Oracle?

**Comment by yhuai** [comment]
Also, I'd like to get the change in master first. Thanks.

**Comment by thomastechs** [comment]
@yhuai  Yes;tested.  So, as you mentioned, will create a new PR against master with the OracleIntegrationSuite.

**Comment by yhuai** [comment]
test this please

**Comment by thomastechs** [comment]
@yhuai   Can the below line wrap pass the style check?
    assert(oracleDialect.getJDBCType(StringType)
      .map(_.databaseTypeDefinition).get == "VARCHAR2(255)")

**Comment by thomastechs** [comment]
Read it as 2 spaces before .map

**Comment by yhuai** [comment]
If you look at the link, you will see (https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50199/consoleFull)
```
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:450: File line length exceeds 100 characters
[error] (sql/test:scalastyle) errors exist
[error] Total time: 7 s, completed Jan 27, 2016 10:12:08 AM
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:450: File line length exceeds 100 characters
[error] (sql/test:scalastyle) errors exist
[error] Total time: 6 s, completed Jan 27, 2016 10:12:47 AM
[error] Got a return code of 1 on line 133 of the run-tests script.
```

Also, I do not think the test in this PR will work. Please put that in the OracleIntegrationSuite in the PR for master.

**Comment by thomastechs** [comment]
@yhuai  working on the OracleIntgerationSuite  (In the process of docker build using Oracle linux)

**Comment by thomastechs** [comment]
@yhuai  and @JoshRosen 
Sorry for the delay. I was working with @jayadevanmurali to bring up the docker based test suite for master branch.
We are able to create the test case with OracleIntegrationSuite and successfully tested this fix in our local environment.The version we used is Oracle 11g:, and used the docker image from wnameless/oracle-xe-11g.
We tried to build using maven repository for the ojdbc jar required for this fix.Added the dependency like below in the pom file (docker integration project).
   <dependency>
            <groupId>com.oracle</groupId>
            <artifactId>ojdbc6</artifactId>
            <version>11.2.0.2.0</version>
           <scope>test</scope>
    </dependency>
But there was no corresponding ojdbc jar in the maven repository. So we had to manually download it to the local and build it.
So, if we create a pull request, how these jar to be uploaded?

**Comment by thomastechs** [comment]
@yhuai 
Created the PR https://github.com/apache/spark/pull/11306/files for the master branch

**Comment by thomastechs** [comment]
@yhuai    The style check error for this PR has been fixed. This is for branch 1.4

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -445,4 +445,10 @@ class JDBCSuite extends SparkFunSuite with BeforeAndAfter {
     assert(agg.getCatalystType(1, "", 1, null) == Some(StringType))
   }
 
+  test("OracleDialect type mapping") {
+    val oracleDialect = JdbcDialects.get("jdbc:oracle://127.0.0.1/db")
+    assert(oracleDialect.getJDBCType(StringType).
+      map(_.databaseTypeDefinition).get == "VARCHAR2(255)")
+  }
+
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: This test does not work, right? If so, can we remove it?

Comment (thomastechs) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: This is similar to testcase written for MySQL,for branch 1.4; which is not actually spinning up any MySQL or Oracle instance; but for the test suite purpose, as I believe.  So far, builds for this PR for branch 1.4, failed due to style checks.

**Comment by yhuai** [comment]
test this please

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 3]
```diff
@@ -445,4 +445,9 @@ class JDBCSuite extends SparkFunSuite with BeforeAndAfter {
     assert(agg.getCatalystType(1, "", 1, null) == Some(StringType))
   }
 
+  test("OracleDialect type mapping") {
+    val oracleDialect = JdbcDialects.get("jdbc:oracle://127.0.0.1/db")
+    assert(oracleDialect.getJDBCType(StringType).
+      map(_.databaseTypeDefinition).get == "VARCHAR2(255)")
+  }
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: OK I see. Originally, I thought it requires a oracle db. Thank you for the explanation.

Then, we should also add this test in the master. Can you submit a PR? We can reuse the jira number.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 4]
```diff
@@ -445,4 +445,9 @@ class JDBCSuite extends SparkFunSuite with BeforeAndAfter {
     assert(agg.getCatalystType(1, "", 1, null) == Some(StringType))
   }
 
+  test("OracleDialect type mapping") {
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Also, it will be good to make the test name more specific on what it will test. I am going to change it while merging.

**Comment by yhuai** [comment]
Merging to 1.4 and 1.5

**Comment by yhuai** [comment]
ok I have merged it into branch 1.4 and branch 1.5. Can you close this PR? Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 3 above)
Comment (thomastechs) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Okei @yhuai    So I shall submit another PR, for the same JIRA with the updates in the JDBCSuite.scala in the master branch. I shall also close this PR
